src/mailmap.c

```diff
@@ -14,195 +14,312 @@
 #include "git2/revparse.h"
 #include "git2/sys/commit.h"
 
-struct mailmap_entry {
-	char* to_name;
-	char* to_email;
-	char* from_name;
-	char* from_email;
-};
+/**
+ * Helper type and methods for the mailmap parser
+ */
+typedef struct char_range {
+	const char *p;
+	size_t len;
+} char_range;
+
+static const char_range NULL_RANGE = {0};
+
+/* Split a range at the first instance of 'c'. Returns whether 'c' was found */
+static bool range_split(
+	char_range range,
+	char c,
+	char_range *before,
+	char_range *after)
+{
+	const char *off;
+
+	*before = *after = NULL_RANGE;
+	before->p = range.p;
+	off = memchr(range.p, c, range.len);
+	if (!off) {
+		before->len = range.len;
+		return false;
+	}
+
+	before->len = off - range.p;
+	after->p = off + 1;
+	after->len = (range.p + range.len) - after->p;
+	return true;
+}
+
+/* Trim whitespace from the beginning and end of the range */
+static void range_trim(char_range *range) {
+	while (range->len > 0 && git__isspace(range->p[0])) {
+		++range->p;
+		--range->len;
+	}
+	while (range->len > 0 && git__isspace(range->p[range->len - 1]))
+		--range->len;
+}
+
+/**
+ * If `buf` is not NULL, copies range into it with a '\0', and bumps buf.
+ * If `size` is not NULL, adds the number of bytes to be written to it.
+ * returns a pointer to the copied string, or NULL.
+ */
+static const char *range_copyz(char **buf, size_t *size, char_range src)
+{
+	char *s = NULL;
+	if (src.p == NULL)
+		return NULL;
+
+	if (size)
+		*size += src.len + 1;
+
+	if (buf) {
+		s = *buf;
+		memcpy(s, src.p, src.len);
+		s[src.len] = '\0';
+		*buf += src.len + 1;
+	}
+	return s;
+}
 
 struct git_mailmap {
-	git_vector lines;
+	git_vector entries;
 };
 
-// Returns -1 on failure, length of the string scanned successfully on success,
-// guaranteed to be less that `length`.
-ssize_t parse_name_and_email(
-	const char *line,
-	size_t length,
-	const char** name,
-	size_t* name_len,
-	const char** email,
-	size_t* email_len,
-	bool allow_empty_email)
+/**
+ * Parse a single entry out of a mailmap file.
+ * Advances the `file` range past the parsed entry.
+ */
+static int git_mailmap_parse_single(
+	char_range *file,
+	bool *found,
+	char_range *real_name,
+	char_range *real_email,
+	char_range *replace_name,
+	char_range *replace_email)
 {
-	const char* email_start;
-	const char* email_end;
-	const char* name_start;
-	const char* name_end;
+	char_range line, comment, name_a, email_a, name_b, email_b;
+	bool two_emails = false;
+
+	*found = false;
+	*real_name = NULL_RANGE;
+	*real_email = NULL_RANGE;
+	*replace_name = NULL_RANGE;
+	*replace_email = NULL_RANGE;
+
+	while (file->len) {
```

Comment:
Feels somewhat inconsistent to use `file->len` here but `file->len > 0` everywhere else. But it doesn't matter much I guess.